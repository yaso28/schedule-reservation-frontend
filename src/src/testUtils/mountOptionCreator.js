import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
