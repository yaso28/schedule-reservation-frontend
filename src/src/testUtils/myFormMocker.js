import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
