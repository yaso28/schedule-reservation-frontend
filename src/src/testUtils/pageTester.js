import { mount } from '@vue/test-utils'
import MyPage from '@/components/containers/MyPage'

class PageTester {
  async mount (pageComponent, route, mountOption) {
    await mountOption.router.push(route)
    const pageWrapper = mount(MyPage, mountOption)
    return pageWrapper.findComponent(pageComponent)
  }

  assertTitleTag () {
    expect(document.title).toMatchSnapshot()
  }

  assertMetaTags () {
    expect(document.querySelectorAll('meta[data-vue-meta]')).toMatchSnapshot()
  }
}

export default new PageTester()
