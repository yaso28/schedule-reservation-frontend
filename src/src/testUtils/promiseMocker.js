import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
