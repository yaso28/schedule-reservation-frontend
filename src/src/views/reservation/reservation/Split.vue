<template>
  <div>
    <slot name="h1" />

    <CForm
      v-if="loaded"
      @submit.prevent="save"
    >
      <CCard>
        <CCardBody>
          <dl class="row">
            <dt class="col-sm-3">
              {{ $t('schedule.ymd') }}
            </dt>
            <dd class="col-sm-9">
              <MyDateSpan :value="reservation.schedule.ymd" />
            </dd>
            <dt class="col-sm-3">
              {{ $t('schedule_place.model') }}
            </dt>
            <dd class="col-sm-9">
              {{ reservation.schedule.schedule_place.name }}
            </dd>
            <dt class="col-sm-3">
              {{ $t('schedule.begins_at') }}
            </dt>
            <dd class="col-sm-9">
              {{ reservation.begins_at }}
            </dd>
            <dt class="col-sm-3">
              {{ $t('schedule.ends_at') }}
            </dt>
            <dd class="col-sm-9">
              {{ reservation.ends_at }}
            </dd>
          </dl>

          <MyTime
            :label="$t('reservation.splits_at')"
            :form="form"
            field="splits_at"
          />
        </CCardBody>
        <CCardFooter>
          <CButton
            class="btn-reset"
            color="secondary"
            @click="form.reset()"
          >
            {{ $t("action.reset") }}
          </CButton>
          <CButton
            type="submit"
            color="primary"
          >
            {{ $t('action.save') }}
          </CButton>
        </CCardFooter>
      </CCard>
    </CForm>
  </div>
</template>

<script>
export default {
  name: 'ReservationSplit',
  components: {
    MyDateSpan: require('@/components/MyDateSpan').default,
    MyTime: require('@/components/inputs/MyTime').default
  },
  inheritAttrs: false,
  inject: ['apiService'],
  props: {
    id: {
      type: [Number, String],
      required: true
    }
  },
  data () {
    return {
      form: null,
      reservation: null
    }
  },
  computed: {
    loaded () {
      return !!(this.form && this.reservation)
    }
  },
  watch: {
    id () {
      this.load()
    }
  },
  mounted () {
    this.load()
  },
  methods: {
    async load () {
      try {
        this.$store.commit('processing/beginGet')
        this.reservation = await this.apiService.getReservation(this.id)
        this.form = this.apiService.makeForm({
          splits_at: ''
        })
      } catch (error) {
        if (!this.apiService.isHandledError(error)) {
          throw error
        }
      } finally {
        this.$store.commit('processing/endGet')
      }
    },
    async save () {
      try {
        this.$store.commit('processing/beginPost')
        this.$store.commit('alert/clear')
        await this.apiService.splitReservation(this.id, this.form)
        this.$store.dispatch('alert/saveSuccess')
        this.$router.push({ name: 'reservation.schedule.edit', params: { id: this.reservation.schedule_id } })
      } catch (error) {
        if (!this.apiService.isHandledError(error)) {
          throw error
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
