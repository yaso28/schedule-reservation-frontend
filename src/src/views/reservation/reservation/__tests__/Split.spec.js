import Split from '../Split'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import pageTester from '@/testUtils/pageTester'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/store/modules')
jest.mock('@/models/MyForm')

const reservation = require('@/testUtils/dummyApiData/reservation/get.json')
const id = reservation.id
const makeFormArgs = {
  splits_at: ''
}
const saveResult = require('@/testUtils/dummyApiData/reservation/split.json')

describe('Split', () => {
  let mountOption
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('マウント', () => {
    let wrapper
    let loadError
    let loadSwitch
    beforeEach(async () => {
      loadError = new Error('load')
      loadSwitch = promiseMocker.createSwitch()
      apiService.getReservation = promiseMocker.mock(reservation, loadError, loadSwitch)
      wrapper = await pageTester.mount(Split, { name: 'reservation.reservation.split', params: { id } }, mountOption)
    })

    test('titleタグ', () => {
      pageTester.assertTitleTag()
    })

    test('metaタグ', () => {
      pageTester.assertMetaTags()
    })

    test('Processingカバー表示開始', () => {
      expect(storeModules.processing.mutations.beginGet).toHaveBeenCalled()
    })

    test('予約レコード取得処理を呼出', () => {
      expect(apiService.getReservation).toHaveBeenCalledWith(id)
    })

    test('Processingカバー表示中', () => {
      expect(storeModules.processing.mutations.endGet).not.toHaveBeenCalled()
    })

    describe('予約レコード取得：エラー', () => {
      beforeEach(() => {
        promiseMocker.setSwitch(loadSwitch, false)
        apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
        promiseMocker.flushPendings()
      })

      test('エラー処理', () => {
        expect(apiService.isHandledError).toHaveBeenCalledWith(loadError)
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
      })
    })

    describe('データ取得：成功', () => {
      let form
      beforeEach(() => {
        form = myFormMocker.createMock(makeFormArgs)
        apiService.makeForm = jest.fn().mockReturnValueOnce(form)
        promiseMocker.flushPendings()
      })

      test('フォーム作成', () => {
        expect(apiService.makeForm).toHaveBeenCalledWith(makeFormArgs)
      })

      test('ページ全体表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
      })

      describe('URL変更', () => {
        let newId
        beforeEach(() => {
          newId = id + 1
          apiService.getReservation = promiseMocker.mockResolved(reservation)
          storeModules.processing.mutations.beginGet.mockReset()
          storeModules.processing.mutations.endGet.mockReset()
          wrapper.setProps({ id: newId })
        })

        test('Processingカバー表示開始', () => {
          expect(storeModules.processing.mutations.beginGet).toHaveBeenCalled()
        })

        test('予約レコード取得処理を呼出', () => {
          expect(apiService.getReservation).toHaveBeenCalledWith(newId)
        })

        test('Processingカバー表示中', () => {
          expect(storeModules.processing.mutations.endGet).not.toHaveBeenCalled()
        })
      })

      describe('リセットボタンクリック', () => {
        beforeEach(() => {
          wrapper.find('button.btn-reset').trigger('click')
        })

        test('フォームリセット', () => {
          expect(form.reset).toHaveBeenCalledWith()
        })
      })

      describe('フォーム送信', () => {
        let router
        let saveError
        let saveSwitch
        beforeEach(() => {
          router = mountOption.router
          router.push = jest.fn()
          saveError = new Error('save')
          saveSwitch = promiseMocker.createSwitch()
          apiService.splitReservation = promiseMocker.mock(saveResult, saveError, saveSwitch)
          wrapper.find('form').trigger('submit')
        })

        test('Processingカバー表示開始', () => {
          expect(storeModules.processing.mutations.beginPost).toHaveBeenCalled()
        })

        test('アラートクリア', () => {
          expect(storeModules.alert.mutations.clear).toHaveBeenCalled()
        })

        test('保存処理を呼出', () => {
          expect(apiService.splitReservation).toHaveBeenCalledWith(id, form)
        })

        test('Processingカバー表示中', () => {
          expect(storeModules.processing.mutations.endPost).not.toHaveBeenCalled()
        })

        describe('保存：エラー', () => {
          beforeEach(() => {
            promiseMocker.setSwitch(saveSwitch, false)
            apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
            promiseMocker.flushPendings()
          })

          test('エラー処理', () => {
            expect(apiService.isHandledError).toHaveBeenCalledWith(saveError)
          })

          test('Processingカバー表示終了', () => {
            expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
          })

          test('成功アラートは呼び出さない', () => {
            expect(storeModules.alert.actions.saveSuccess).not.toHaveBeenCalled()
          })

          test('リダイレクトしない', () => {
            expect(router.push).not.toHaveBeenCalled()
          })
        })

        describe('保存：成功', () => {
          beforeEach(() => {
            promiseMocker.flushPendings()
          })

          test('成功アラート', () => {
            expect(storeModules.alert.actions.saveSuccess).toHaveBeenCalled()
          })

          test('リダイレクト', () => {
            expect(router.push).toHaveBeenCalledWith({ name: 'reservation.schedule.edit', params: { id: reservation.schedule_id } })
          })

          test('Processingカバー表示終了', () => {
            expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
          })
        })
      })
    })
  })
})
