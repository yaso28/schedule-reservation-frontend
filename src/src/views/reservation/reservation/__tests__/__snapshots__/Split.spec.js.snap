// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Split マウント metaタグ 1`] = `
NodeList [
  <meta
    content="noindex"
    data-vmid="robots"
    data-vue-meta="2"
    name="robots"
  />,
]
`;

exports[`Split マウント titleタグ 1`] = `"やそスポーツクラブ | 会場予約分割"`;

exports[`Split マウント データ取得：成功 ページ全体表示 1`] = `
<div>
  <h1>
    会場予約分割
  </h1>
   
  <form
    class=""
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <dl
          class="row"
        >
          <dt
            class="col-sm-3"
          >
            
            日付
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            <span>
              
  2020/11/01
  
              <span
                class="text-danger"
              >
                
    (日)
  
              </span>
            </span>
          </dd>
           
          <dt
            class="col-sm-3"
          >
            
            場所
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            
            体育館＠公民館別棟
          
          </dd>
           
          <dt
            class="col-sm-3"
          >
            
            開始時刻
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            
            13:00
          
          </dd>
           
          <dt
            class="col-sm-3"
          >
            
            終了時刻
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            
            16:00
          
          </dd>
        </dl>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="splits_at"
          >
             分割する時刻 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="splits_at"
              step="900"
              type="time"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
       
      <footer
        class="card-footer"
      >
        <button
          class="btn btn-reset btn-secondary"
          type="button"
        >
          
          リセット
        
        </button>
         
        <button
          class="btn btn-primary"
          type="submit"
        >
          
          保存
        
        </button>
      </footer>
    </div>
  </form>
</div>
`;
